proposals/0000-borrow-operator.md

```diff
@@ -0,0 +1,195 @@
+each argument appears to the callee as formally independent values. As an
+example, consider the following code:
+
+```
+func useValueTwice(_ x: Type, _ y: inout Type) {}
+
+var a: Type = ...
+useValueTwice(a, &a)
+```
+
+Formally during evaluation of `useValueTwice`, the compiler will perform an
+immutable exclusive access to `a`'s storage and copy `a` before beginning a
+mutable exclusive access to `a`'s storage. The reason that function evaluation
+is performed in this way is that otherwise, we could suffer from several
+undesireable outcomes, for example:
+
+1. Potentially trigger exclusivity violations. For instance, if we formally
+accessed the inout parameters memory before performing the `pass-by-value`, we
+would have an exclusivity violation when attempting to exclusively access the
+`var` memory to perform our copy.
+
+2. If we did not copy the parameter, we would need to either perform a `take` of
+the non-`inout` parameter meaning that said memory would now be
+invalidated. This would mean that we couldn't pass in var storage directly as an
+`inout` parameter since the memory would be invalidated.
+
+For this and other reasons, we pass these arguments as `pass-by-value`
+formally. These copies are observable program behavior and as shown above are
+desireable in our formal model of callsite evaluation. In certain cases though,
+this `pass-by-value` behavior and the implicit copies it requires the compiler
+to insert may be undesireable due to the author needing strict performance
+guarantees that implicit copies like the above are never generated. For most
+programs this guarantee is not needed since the optimizer generally does a good
+enough job eliminating such copies. But in critical system code, "good enough"
+is not an acceptable constraint and a true semantic language level guarantee is
+needed.
+
+## Proposed solution
+
+We propose introducing a new context sensitive keyword called `borrow` that when
+applied to a local var, computed var, or global var passed as a function
+argument forces Swift to use a `pass-by-immutable-borrow` convention on a
+specific argument instead of a `pass-by-value` convention. This would forbid the
+compiler from inserting such copies and instead cause the compiler to pass the
+argument in the same manner as an inout argument except performing an immutable
+exclusive access to memory instead of a mutable exclusive access. The result of
+the usage of borrow is that a performance minded programmer would be able to
+sleep well at night knowing that their code doesn't have copies in this code.
+
+## Detailed design
+
+The `borrow` keyword can be applied in argument positions to local, computed,
+and global `vars`. It will cause the `var` to use a `pass-by-immutable-borrow`
+convention. Lets revisit our `useValueTwice` example from above in more detail:
+
+```
+func useValueTwice(_ x: Type, _ y: inout Type) {}
+
+var a: Type = ...
+useValueTwice(a, &a)
+```
```

Comment:
I am very confused at this point.  This is exactly the same code as above, but the discussion below makes it sound like you are proposing to _change_ the semantics of this code.

Shouldn't there be an example here that actually has `borrow` keywords in it?

Reply:
I am explaining to the reader in the detailed design section in detail the way the callsite evaluation occurs in order to contrast it with when we apply the borrow convention. I don't think that I can assume that all of our users understand the way that the language defines formal evaluation of call sites.

Reply:
I think I have an idea of how to make this a little clearer. I should first define the rules for formal evaluation of callsites and then show the example afterwards.